File: quicksort.py
import random
"""
Ejercicios hechos por:
    CALEB SANCHEZ SOLORZANO
    SEBASTIAN SEGURA MENDEZ
    ANGELLY CORTES HERRERA
    LUIS QUESADA BARRANTES
    ESTEBAN RAMIREZ GARCIA
"""

def exchange(arr, i, j):
    temp = arr[i]
    arr[i] = arr[j]
    arr[j] = temp


def partition(arr, low, high):
    pivot = arr[low]
    i = low
    j = high - 1
    while i < j:
        while i < j:
            if arr[i] <= pivot:
                i += 1
            else:
                break

        while True:
            if arr[j] > pivot:
                j -= 1
            else:
                break
        if i < j:
            exchange(arr, i, j)
    exchange(arr, j, low)
    return j


def quicksort1(arr, low, high):
    if low >= high: return
    j = partition(arr, low, high)
    quicksort1(arr, low, j)
    quicksort1(arr, j + 1, high)


def quicksort(arr):
    quicksort1(arr, 0, len(arr))


def check(arr):
    return arr == sorted(arr)


if __name__ == '__main__':
    length = 25
    a = [random.randint(1, 100000) for i in range(length)]
    print("Array antes de quicksort:", a)
    quicksort(a)
    print("Array despues de quicksort:", a)
    print("La lista está ordenada?", "si" if check(a) else "no")

File: quicksort.cpp
/*
Ejercicios hechos por:

CALEB SANCHEZ SOLORZANO
SEBASTIAN SEGURA MENDEZ
ANGELLY CORTES HERRERA
LUIS QUESADA BARRANTES
ESTEBAN RAMIREZ GARCIA

*/


#include <iostream>
#include <cstdlib>
#include <ctime>

using namespace std;

int partition(int arr[], int low, int high);

void quicksort(int arr[], int l);

void quicksort(int arr[], int low, int high);

void exchange(int arr[], int i, int j);

int random(int min, int max);

void print(int arr[], int len);

int main() {
    int max = 100000;
    int length = 10;
    int a[length];
    for (int i = 0; i < length; ++i) {
        a[i] = random(0, max);
    }
    cout << "Array antes:";
    print(a, length);
    quicksort(a, length);
    cout << "Array despues:";
    print(a, length);
    return 0;
}

void print(int arr[], int length) {
    cout << "\n";
    for (int i = 0; i < length; i++) {
        cout << arr[i] << ((i + 1 == length) ? "" : ",");
    }
    cout << "\n";
}

void quicksort(int arr[], int length) {
    quicksort(arr, 0, length);
}

void quicksort(int arr[], int low, int high) {
    if (low >= high) return;
    int p = partition(arr, low, high);
    quicksort(arr, low, p);
    quicksort(arr, p + 1, high);
}

int partition(int arr[], int low, int high) {
    int pivot = arr[low];
    int i = low;
    int j = high;
    while (i < j) {//n-1
        while (arr[++i] <= pivot);//n/k
        while (arr[--j] > pivot);//n/k

        if (i < j)
            exchange(arr, i, j);
    }
    exchange(arr, j, low);
    return j;
}

void exchange(int arr[], int i, int j) {
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

int random(int min, int max) { //https://stackoverflow.com/a/7560171/11449132
    static bool first = true;
    if (first) {
        srand(time(NULL)); //seeding for the first time only!
        first = false;
    }
    return min + rand() % ((max + 1) - min);
}